src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs

```diff
@@ -147,83 +156,118 @@ static void RegisterImplActions(CompilationStartAnalysisContext compilationConte
-                {
-                    string text = line.ToString();
-                    if (string.IsNullOrWhiteSpace(text))
-                    {
-                        continue;
-                    }
+            var additionalFileInfo = new AdditionalFileInfo(path, sourceText, isShippedApi);
 
-                    rank++;
+            foreach (var line in sourceText.Lines)
+            {
+                // Skip whitespace.
+                var text = line.ToString();
+                if (string.IsNullOrWhiteSpace(text))
+                    continue;
 
-                    if (text == NullableEnable)
-                    {
-                        maxNullableRank = Math.Max(rank, maxNullableRank);
-                        continue;
-                    }
+                lineNumber++;
 
-                    var apiLine = new ApiLine(text, line.Span, sourceText, path, isShippedApi);
-                    if (text.StartsWith(RemovedApiPrefix, StringComparison.Ordinal))
-                    {
-                        string removedtext = text[RemovedApiPrefix.Length..];
-                        removedBuilder.Add(new RemovedApiLine(removedtext, apiLine));
-                    }
-                    else
-                    {
-                        apiBuilder.Add(apiLine);
-                    }
+                if (text == NullableEnable)
+                {
+                    lastNullableLineNumber = lineNumber;
+                    continue;
+                }
+
+                var apiLine = new ApiLine(text, line.Span, additionalFileInfo);
+                if (text.StartsWith(RemovedApiPrefix, StringComparison.Ordinal))
+                {
+                    var removedText = text[RemovedApiPrefix.Length..];
+                    removedBuilder.Add(new RemovedApiLine(removedText, apiLine));
+                }
+                else
+                {
+                    apiBuilder.Add(apiLine);
                 }
             }
 
-            return new ApiData(apiBuilder.ToImmutable(), removedBuilder.ToImmutable(), maxNullableRank);
+            return new ApiData(apiBuilder.ToImmutableAndFree(), removedBuilder.ToImmutableAndFree(), lastNullableLineNumber);
         }
 
-        private static bool TryGetApiData(AnalyzerOptions analyzerOptions, Compilation compilation, bool isPublic, List<Diagnostic> errors, CancellationToken cancellationToken, out ApiData shippedData, out ApiData unshippedData)
+        private static bool TryGetApiData(AnalyzerOptions analyzerOptions, Compilation compilation, bool isPublic, List<Diagnostic> errors, CancellationToken cancellationToken, [NotNullWhen(true)] out ApiData? shippedData, [NotNullWhen(true)] out ApiData? unshippedData)
         {
-            if (!TryGetApiText(analyzerOptions.AdditionalFiles, isPublic, cancellationToken, out var shippedText, out var unshippedText))
+            using var allShippedData = ArrayBuilder<ApiData>.GetInstance();
```

Comment:
>using var allShippedData [](http://example.com/codeflow?start=12&length=24)

This threw me for a loop until I saw this was coming from an implementation in this repo. I assume we don't want a similar disposable pattern for the one in Roslyn?

Reply:
The Roslyn one is disposable as well, just a slightly different pattern (due to compiler team concerns)

Reply:
Weird, I'm not seeing the Roslyn ones as IDisposable from code inspection, but I'm probably missing it

Reply:
we have a struct disposable wrapper around ours.  look for `using var _ = ArrayBuilder...` in roslyn.sln :)